eis_toolkit/prediction/cnn_classification_and_probability.py

```diff
@@ -0,0 +1,417 @@
+from typing import Literal, Union
+
+import numpy as np
+import pandas as pd
+import sklearn.preprocessing
+import tensorflow as tf
+from beartype import beartype
+from keras import Model
+from pandas import DataFrame
+from sklearn.metrics import confusion_matrix
+from sklearn.preprocessing import OneHotEncoder, StandardScaler
+from sklearn.utils.class_weight import compute_sample_weight
+
+from eis_toolkit.exceptions import CNNException, CNNRunningParameterException, InvalidArgumentTypeException
+from eis_toolkit.prediction.model_performance_estimation import performance_model_estimation
+
+
+def normalize_the_data(scaler_agent: sklearn.preprocessing, data: np.ndarray):
+    """
+    Normalize multi-dimensional data using a specified scaler.
+    
+    This function applies a scaling technique to each feature in the data, transforming the data to a specified range or distribution. It is particularly useful for normalizing image data or similarly structured multi-dimensional data.
+
+    Parameters:
+        - scaler_agent: An instance of a preprocessing scaler from sklearn.preprocessing (e.g., MinMaxScaler, StandardScaler) that will be used to apply normalization.
+        - data: A NumPy ndarray containing the data to be normalized. The data is expected to be in a multi-dimensional format, typically including dimensions for sample size, height, width, and channels (e.g., for image data).
+
+    Returns:
+        - normalized_data: The input data normalized according to the scaler_agent. The output data maintains the original shape of the input data.
+        
+    """
+    number_of_sample, h, w, c = data.shape
+    temp = scaler_agent.transform(data.reshape(-1, data.shape[-1]))
+    normalized_data = temp.reshape(number_of_sample, h, w, c)
+    return normalized_data
+
+
+@beartype
+def make_one_hot_encoding(labels: np.ndarray):
+     """
+    Convert categorical labels to a one-hot encoded format.
+
+    This function transforms a list of numerical or categorical labels into a one-hot encoded matrix, which is a binary matrix representing the categorization of each label. One-hot encoding is commonly used in machine learning to handle categorical features.
+
+    Parameters:
+        - labels: A NumPy ndarray containing the labels to be encoded. Each label corresponds to a category or class.
+
+    Returns:
+        - label_encoded: A matrix where each row corresponds to a one-hot encoded representation of the input labels.
+
+    Raises:
+        - InvalidArgumentTypeException: Raised if the input 'labels' is None, indicating that no labels were provided for encoding.
+
+    """
+    if labels is None:
+        raise InvalidArgumentTypeException
+
+    # to categorical
+    enc = OneHotEncoder(handle_unknown="ignore")
+    # train and valid set
+    temp = np.reshape(labels, (-1, 1))
+    label_encoded = enc.fit_transform(temp).toarray()
+    return label_encoded
+
+
+@beartype
+def do_the_cnn(
```

Comment:
This is not a good name for a function. And mark it as private (by prefexing it with "_") as users are not meant to import it.

Reply:
ok I marked as private

Reply:
I gave better name